### PR TITLE
refactor: migrate test support to ts

### PR DESCRIPTION
**Pull request by ephys**
### Pull Request Checklist

- [ ] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

This PR migrates `test/supports.js` and its dependencies to TS. Some of these changes were initially done in https://github.com/sequelize/sequelize/pull/14280, but they're often needed in new tests (such as the ones being written in https://github.com/sequelize/sequelize/tree/ephys/fix-13817), so this PR extracts them out of that PR

### CODE REVIEW

File: test/support.ts [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+
+  if (onNext || unhandledRejections) {
+    return;
+  }
+
+  console.error('An unhandled rejection occurred:');
+  throw e;
+});
+
+afterEach(() => {
```
Comment (WikiRik) [inline test/support.ts:62]: Why did you remove the check for global.afterEach?

File: test/support.ts [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+ * during this test (instead of failing the test)
+ */
+export async function nextUnhandledRejection() {
+  return new Promise((resolve, reject) => {
+    onNextUnhandledRejection = reject;
+  });
+}
+
+/**
+   * Pushes all unhandled rejections that occur during this test onto destArray
```
Comment (WikiRik) [inline test/support.ts:78]: I think this indentation is off

File: test/support.ts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+  return sequelize;
+}
+
+export function createSequelizeInstance(options: Options = {}) {
+  options.dialect = getTestDialect();
+
+  const config = Config[options.dialect];
+
+  const sequelizeOptions = defaults(options, {
```
Comment (WikiRik) [inline test/support.ts:123]: Why is `odbcConnectionString` not mentioned here anymore?

File: test/support.ts [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+      return x;
+    }
+  }
+
+  // @ts-expect-error
+  return new ModdedQueryGenerator({ sequelize, _dialect: sequelize.dialect });
+}
+
+export function getTestDialect(): Dialect {
+  let envDialect = process.env.DIALECT || 'mysql';
```
Comment (WikiRik) [inline test/support.ts:212]: Why would we default to mysql here? I know it's in the original file, but seems odd to me

File: test/support.ts [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+export function expectsql(
+  query: { query: string, bind: unknown } | Error,
+  assertions: { query: PartialRecord<ExpectationKey, string | Error>, bind: PartialRecord<ExpectationKey, unknown> },
+): void;
+export function expectsql(
+  query: string | Error,
+  assertions: PartialRecord<ExpectationKey, string | Error>,
+): void;
+export function expectsql(
```
Comment (WikiRik) [inline test/support.ts:250]: Why are we exporting this multiple times?

File: test/support.ts [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+export function rand() {
+  return Math.floor(Math.random() * 10e5);
+}
+
+export function isDeepEqualToOneOf(actual: unknown, expectedOptions: unknown[]): boolean {
+  return expectedOptions.some(expected => isDeepStrictEqual(actual, expected));
+}
+
+/**
+   * Reduces insignificant whitespace from SQL string.
```
Comment (WikiRik) [inline test/support.ts:299]: Indentation is off

File: test/support.ts [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+   *
+   * @param sql the SQL string
+   * @returns the SQL string with insignificant whitespace removed.
+   */
+export function minifySql(sql: string): string {
+  // replace all consecutive whitespaces with a single plain space character
+  return sql.replace(/\s+/g, ' ')
+  // remove space before coma
```
Comment (WikiRik) [inline test/support.ts:307]:
```suggestion
  // remove space before comma
```

File: test/support.ts [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+    .replace(/ ,/g, ',')
+  // remove whitespace at start & end
+    .trim();
+}
+
+export const sequelize = createSequelizeInstance();
+
+before(function onBefore() {
```
Comment (WikiRik) [inline test/support.ts:315]: Why is the check for global.beforeEach removed here?

**Review by WikiRik — COMMENTED** [review]
Few remarks/questions

File: test/support.ts (hunk 3 above)
Comment (ephys) [reply test/support.ts:123]: It wasn't correctly set (according to the typings, and its usage in ibmi's connection manager). The `obdcConnectionString` isn't a valid option, it's named `dialectOptions.obdcConnectionString`. Which means it was unused

File: test/support.ts (hunk 4 above)
Comment (ephys) [reply test/support.ts:212]: Sounds weird to me too. I think we should throw if DIALECT is not set

File: test/support.ts (hunk 5 above)
Comment (ephys) [reply test/support.ts:250]: It's overloading (https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads)

Comment (WikiRik) [reply test/support.ts:250]: Ah yes, I was looking for the term of this already. But the thing I am missing here is why we use this here, because we have 2 arguments in all three options. And the third just looks like a merger of the first two. But probably I'm missing something here in my thought process

Comment (ephys) [reply test/support.ts:250]: The implementation is always a merger of all previous overloads, but it's 'internal' of sorts (even if exported), TypeScript won't include it as part of the possible overloads when you use the method

File: test/support.ts (hunk 1 above)
Comment (ephys) [reply test/support.ts:62]: All of our tests are run by mocha, so these are always defined

**Review by WikiRik — APPROVED** [review]
